src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java

```diff
@@ -0,0 +1,281 @@
+    }
+
+    /**
+     * Returns the how much space an index is taking up in the cache as a percentage of the total cache capacity.
+     *
+     * @param indexName name of the index
+     * @return Percentage of the cache full
+     */
+    public Float getWeightForIndexAsPercentage(final String indexName) {
+        return 100 * getWeightForIndexInKilobytes(indexName) / (float) KNNSettings.getCircuitBreakerLimit().getKb();
+    }
+
+    /**
+     * Getter for maximum weight of the cache.
+     *
+     * @return maximum cache weight
+     */
+    public long getCacheMaxWeightInKb() {
+        return maxWeight;
+    }
+
+    /**
+     * Getter for cache stats.
+     *
+     * @return cache status
+     */
+    public CacheStats getCacheStats() {
+        return cache.stats();
+    }
+
+    /**
+     * Getter for cache as a map.
+     *
+     * @return cache as a map
+     */
+    public Map<String, NativeMemoryAllocation> getCacheAsMap() {
+        return cache.asMap();
+    }
+
+    /**
+     * Retrieves NativeMemoryAllocation associated with the nativeMemoryEntryContext.
+     *
+     * @param nativeMemoryEntryContext Context from which to get NativeMemoryAllocation
+     * @param isAbleToTriggerEviction Determines if getting this allocation can evict other entries
+     * @return NativeMemoryAllocation associated with nativeMemoryEntryContext
+     * @throws ExecutionException if there is an exception when loading from the cache
+     */
+    public NativeMemoryAllocation get(NativeMemoryEntryContext<?> nativeMemoryEntryContext,
+                                      boolean isAbleToTriggerEviction) throws ExecutionException {
+        if (!isAbleToTriggerEviction &&
+                !cache.asMap().containsKey(nativeMemoryEntryContext.getKey()) &&
+                maxWeight - getCacheWeightInKilobytes() - nativeMemoryEntryContext.calculateSizeInKb() <= 0
+        ) {
+            throw new NativeMemoryThrottleException("Failed to load \"" + nativeMemoryEntryContext.getKey() +
+                    "\" into memory.");
+        }
+
+        return cache.get(nativeMemoryEntryContext.getKey(), nativeMemoryEntryContext::load);
+    }
+
+    /**
+     * Invalidate entry from the cache.
+     *
+     * @param key Identifier of entry to invalidate
+     */
+    public void invalidate(String key) {
+        cache.invalidate(key);
+    }
+
+    /**
+     * Invalidate all entries in the cache.
+     */
+    public void invalidateAll() {
+        cache.invalidateAll();
+    }
+
+    /**
+     * Returns whether or not the capacity of the cache has been reached
+     *
+     * @return Boolean of whether cache limit has been reached
+     */
+    public Boolean isCacheCapacityReached() {
+        return cacheCapacityReached.get();
+    }
+
+    /**
+     * Sets cache capacity reached
+     *
+     * @param value Boolean value to set cache Capacity Reached to
+     */
+    public void setCacheCapacityReached(Boolean value) {
```

Comment:
No, it needs to be public because it needs to be unset here: https://github.com/jmazanec15/k-NN-1/blob/cache-refactor/src/main/java/org/opensearch/knn/index/KNNCircuitBreaker.java#L91